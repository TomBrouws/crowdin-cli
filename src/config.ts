import fs from "fs";
import shell from "shelljs";

import log from "./utils/logging";

import { Config } from "./types";

if (!process.env.CROWDIN_KEY || !process.env.CROWDIN_PROJECT_NAME) {
  log.error("Please set your Crowdin key and project name in your .env file.");
  process.exit(1);
}

const { stdout } = shell.exec("git rev-parse --abbrev-ref HEAD | tr / -", {
  silent: true
});
const branch =
  process.env.NODE_ENV !== "test" ? stdout.replace("\n", "") : "test-branch";

const PROJECT_DIR = process.cwd();

const WORKING_DIR =
  process.env.NODE_ENV !== "test" ? process.cwd() : `${process.cwd()}/tests`;

const INTL_DIR = `${WORKING_DIR}/intl`;
const MESSAGES_DIR = `${WORKING_DIR}/messages`;

const config: Config = {
  BIN: `${PROJECT_DIR}/node_modules/.bin`,
  BRANCH: branch,
  CROWDIN_KEY: process.env.CROWDIN_KEY || "",
  CROWDIN_PROJECT_NAME: process.env.CROWDIN_PROJECT_NAME || "",
  INTL_DIR,
  MESSAGES_DIR,
  MESSAGES_PATTERN: `${MESSAGES_DIR}/**/*.json`,
  NODE_EXEC: process.execPath,
  TRANSLATIONS_DIR: `${WORKING_DIR}/src/intl`,
  TRANSLATIONS_FILE: `${INTL_DIR}/english.source.json`
};

if (!fs.existsSync(`${config.BIN}/formatjs`)) {
  log.error("@formatjs/cli executable not found in project directory.");
  process.exit(1);
}

export default config;
