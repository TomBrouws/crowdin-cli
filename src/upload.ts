import fs from "fs";
import {
  createFile,
  updateOrRestoreFile,
  createBranch,
  isCommonErrorResponse,
  unwrapErrorResponse,
} from "./lib/crowdin";
import log from "./utils/logging";
import chalk from "chalk";
import {
  CommonErrorResponse,
  ValidationErrorResponse,
} from "@crowdin/crowdin-api-client";

interface UploadOptions {
  translationsFile: string;
  branchName: string;
  clearOnUpdate?: boolean;
}

export default async (
  options: UploadOptions
): Promise<{
  fileId: number;
} | void> => {
  log.info("Uploading source file to Crowdin");

  const file = fs.createReadStream(options.translationsFile);

  try {
    const branchResponse = await createBranch(options.branchName);

    if (isCommonErrorResponse(branchResponse)) {
      log.error(branchResponse.error.message);
    } else {
      const fileResponse = await createFile(branchResponse.data.name, file);
      log.success(`Created branch ${chalk.bold(branchResponse.data.name)}`);
      log.success(
        `Uploaded source file (id: ${
          fileResponse.data.id
        }) to branch: ${chalk.bold(branchResponse.data.name)}`
      );

      return {
        fileId: fileResponse.data.id,
      };
    }
  } catch (errorResponse) {
    const error = unwrapErrorResponse(
      errorResponse as CommonErrorResponse | ValidationErrorResponse
    );

    if (error.code === "notUnique") {
      // Branch already exists. Update the file.
      log.info(`Branch ${chalk.bold(options.branchName)} already exists`);
      log.info(
        `Updating source file in branch: ${chalk.bold(options.branchName)}`
      );

      try {
        const fileResponse = await updateOrRestoreFile({
          branchName: options.branchName,
          clearTranslationsAndApprovals: options.clearOnUpdate,
          file,
        });

        if (isCommonErrorResponse(fileResponse)) {
          log.error(
            `${fileResponse.error.code}: ${fileResponse.error.message}`
          );

          return;
        }

        log.success("Source file updated");

        return {
          fileId: fileResponse.data.id,
        };
      } catch (error) {
        typeof error === "string" ? log.error(error) : console.log(error);
      }
    } else {
      // Something else went wrong. Sometimes after a branch is
      // deleted in Crowdin, it takes a couple seconds until a new one
      // can be created. Itâ€™s probably that.
      log.error(error.code + ": " + error.message);
    }
  }
};
