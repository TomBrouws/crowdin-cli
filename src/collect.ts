import { sync as globSync } from "glob";
import { sync as mkdirpSync } from "mkdirp";
import fs from "fs";
import shell from "shelljs";

import config from "./config";

import sortByKey from "./utils/sortByKey";
import log from "./utils/logging";

import { Messages, Descriptor } from "./types";

export default async (glob: string) => {
  mkdirpSync(config.INTL_DIR);

  log.info("", "Collecting messages");

  const files = globSync(glob);

  const cmd = [
    config.NODE_EXEC,
    `${config.BIN}/formatjs`,
    "extract",
    files.join(" "),
    `--messages-dir=${config.MESSAGES_DIR}`
  ];

  const { stderr } = shell.exec(cmd.join(" "));

  if (stderr) {
    log.error(stderr);
    process.exit(1);
  }

  const messages: Messages = globSync(config.MESSAGES_PATTERN)
    .map(filename => fs.readFileSync(filename, "utf8"))
    .map(file => JSON.parse(file))
    .reduce((collection, descriptors) => {
      try {
        descriptors.forEach(
          ({ id, defaultMessage, description }: Descriptor) => {
            if (collection[id] && collection[id].message !== defaultMessage) {
              log.error(`Duplicate message id: ${id}`);
              process.exit(1);
            }

            if (!defaultMessage) {
              log.error(`No default message supplied for: ${id}`);
              process.exit(1);
            }

            // Chrome JSON format
            collection[id] = { message: defaultMessage, description };
          }
        );
      } catch (error) {
        log.error(error);
        process.exit(1);
      }
      return collection;
    }, {});

  const sortedMessages = sortByKey(messages);

  log.info("", `${Object.keys(sortedMessages).length} translations found.`);

  const newTranslations = JSON.stringify(sortedMessages, null, 2);

  log.info("", "Writing new translations source file.");
  fs.writeFileSync(config.TRANSLATIONS_FILE, newTranslations);

  return;
};
