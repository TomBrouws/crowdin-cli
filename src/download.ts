import { updateFile, exportFile } from './utils/client';
import sanitizeHtml from 'sanitize-html';
import { sync as globSync } from 'glob';
import { sync as mkdirpSync } from 'mkdirp';
import fs from 'fs';

import shell from 'shelljs';

import config, { LocaleLanguageMap, Languages, Locales } from './config';
import log from './utils/logging';

export default async () => {
  /* sync source file to prevent branching issues */

  if (!fs.existsSync(config.TRANSLATIONS_FILE)) {
    log.error('Source json file does not exist.');
    process.exit(1);
  }

  log.info('CROWDIN', 'Starting download process');

  mkdirpSync(config.TRANSLATIONS_DIR);

  const updateResponse = await updateFile(config.BRANCH);

  log.info('', 'Syncing source.json');

  if (!updateResponse?.data?.success) {
    const { error } = updateResponse.data;
    log.error(error?.message || `Something went wrong when uploading source.json`);
    process.exit(1);
  }

  log.info('', 'Downloading translations')

  const exportFileResponses = await Promise.all(
    Object.keys(LocaleLanguageMap).map(
      (locale) => exportFile(config.BRANCH, LocaleLanguageMap[(locale as Locales)])
    )
  );

  log.info('', `Writing translations to: ${config.TRANSLATIONS_DIR}`);

  await Promise.all(
    Object.keys(LocaleLanguageMap).map((locale, i) => {
      const translations = exportFileResponses[i].data;
      // @ts-ignore
      const destination = `${config.TRANSLATIONS_DIR}/${locale}.js`;
      const jsData = `// Auto generated file. Do no change. Go to Crowdin to update the translations and run './node_modules/.bin/mollie-crowdin download' to update this file.\nexport default ${JSON.stringify(translations)};`;
      return new Promise(
        (resolve, reject) => fs.writeFile(destination, jsData, () => resolve(true))
      );
    })
  );

  log.info('', 'Running prettier');

  const prettier = `${config.BIN}/prettier`;

  shell.exec([
    prettier,
    '--loglevel silent',
    `--write "${config.TRANSLATIONS_DIR}/*.+(js)"`,
  ].join(' '));

  log.success('Updated translations ðŸŽ‰');

}