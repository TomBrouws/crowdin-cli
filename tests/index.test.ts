import fs from "fs";
import CrowdinApiClient, { Credentials } from "@crowdin/crowdin-api-client";
import expectedJson from "./fixtures/expected.json";
import config from "../src/config";
import program from "../src/cli";
import collect from "../src/collect";
import download from "../src/download";
import upload from "../src/upload";
import deleteBranch from "../src/delete-branch";

jest.mock("@crowdin/crowdin-api-client", () =>
  jest.fn().mockReturnValue({
    uploadStorageApi: {
      addStorage: jest.fn().mockResolvedValue({
        data: { id: 324234 },
      }),
    },
    translationsApi: {
      exportProjectTranslation: jest.fn().mockResolvedValue({
        data: { url: "https://example.com/mock.json" },
      }),
    },
    sourceFilesApi: {
      listProjectBranches: jest.fn().mockResolvedValue({
        data: [{ data: { id: 3235234 } }],
      }),
      listProjectFiles: jest.fn().mockResolvedValue({
        data: [{ data: { id: 234234 } }],
      }),
      updateOrRestoreFile: jest.fn().mockResolvedValue({
        data: { id: 234234 },
      }),
      createBranch: jest.fn().mockResolvedValue({
        data: { id: 234234, name: "mock-branch-name" },
      }),
      createFile: jest.fn().mockResolvedValue({
        data: { id: 234234 },
      }),
    },
  })
);

jest.mock(
  "axios",
  jest.fn().mockReturnValue({
    get: jest.fn().mockResolvedValue({
      statusText: "OK",
      data: {
        "mollie-crowdin-content": { message: "Dit is een test" },
        "mollie-crowdin-title": { message: "Hi" },
      },
    }),
  })
);

jest.mock(
  "fs",
  jest.fn().mockReturnValue({
    ...jest.requireActual("fs"),
    createReadStream: jest.fn().mockReturnValue("mockReadStream"),
  })
);

jest.mock("../src/collect", () => {
  return {
    __esModule: true,
    default: jest.fn(jest.requireActual("../src/collect").default),
  };
});

jest.mock("../src/upload", () => {
  return {
    __esModule: true,
    default: jest.fn(jest.requireActual("../src/upload").default),
  };
});

jest.mock("../src/download", () => {
  return {
    __esModule: true,
    default: jest.fn(jest.requireActual("../src/download").default),
  };
});

jest.mock("../src/delete-branch", () => {
  return {
    __esModule: true,
    default: jest.fn(jest.requireActual("../src/delete-branch").default),
  };
});

const mockGlob = "tests/fixtures/**/*.ts*";

describe("CLI", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("correctly handles `collect` command", async () => {
    await program(["node", "test", "collect", mockGlob]);
    expect(collect).toHaveBeenCalledTimes(1);
    expect(collect).toHaveBeenCalledWith(mockGlob);
  });

  it("correctly handles `upload` command", async () => {
    await program(["node", "test", "upload", mockGlob]);
    expect(collect).toHaveBeenCalledWith(mockGlob);
    await program([
      "node",
      "test",
      "upload",
      mockGlob,
      "-b",
      "custom-branch-name",
    ]);
    expect(upload).toHaveBeenCalledWith(
      expect.objectContaining({
        branchName: "custom-branch-name",
      })
    );
    expect(collect).toHaveBeenCalledTimes(2);
    expect(upload).toHaveBeenCalledTimes(2);
  });

  it("correctly handles `download` command", async () => {
    await program(["node", "test", "download"]);
    expect(download).toHaveBeenCalledWith(
      expect.objectContaining({ typescript: false })
    );
    await program(["node", "test", "download", "-b", "custom-branch-name"]);
    expect(download).toHaveBeenCalledWith(
      expect.objectContaining({
        typescript: false,
        branchName: "custom-branch-name",
      })
    );
    await program(["node", "test", "download", "--typescript"]);
    expect(download).toHaveBeenCalledWith(
      expect.objectContaining({ typescript: true })
    );
    expect(download).toHaveBeenCalledTimes(3);
  });

  it("correctly handles `delete-branch` command", async () => {
    await program(["node", "test", "delete-branch"]);
    expect(deleteBranch).toHaveBeenCalledWith(
      expect.objectContaining({
        branchName: config.BRANCH_NAME,
      })
    );
    await program([
      "node",
      "test",
      "delete-branch",
      "--branch-name",
      "custom-branch-name",
    ]);
    expect(deleteBranch).toHaveBeenCalledWith(
      expect.objectContaining({
        branchName: "custom-branch-name",
      })
    );
    expect(deleteBranch).toHaveBeenCalledTimes(2);
  });
});

describe("Handlers", () => {
  beforeEach(() => {
    // Clean up auto-generated directories
    fs.rmdirSync(config.INTL_DIR, { recursive: true });
    fs.rmdirSync(config.TRANSLATIONS_DIR, { recursive: true });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("collects messages correctly", async () => {
    expect.assertions(2);

    // Collects all messages from a component and creates english.source.json
    await collect(mockGlob);
    expect(
      fs.existsSync(`${config.INTL_DIR}/english.source.json`)
    ).toBeTruthy();

    // english.source.json matches the expected JSON file
    const sourceJson = require(`${config.INTL_DIR}/english.source.json`);
    expect(sourceJson).toEqual(expectedJson);
  });

  it("creates a branch on Crowdin and uploads the collected messages", async () => {
    expect.assertions(6);

    const credentials: Credentials = {
      token: config.CROWDIN_PERSONAL_ACCESS_TOKEN,
    };
    const { sourceFilesApi, uploadStorageApi } = new CrowdinApiClient(
      credentials
    );

    await upload({
      translationsFile: config.TRANSLATIONS_FILE,
      branchName: config.BRANCH_NAME,
    });

    expect(sourceFilesApi.createBranch).toHaveBeenCalledTimes(1);
    expect(sourceFilesApi.createBranch).toHaveBeenCalledWith(
      config.CROWDIN_PROJECT_ID,
      {
        name: config.BRANCH_NAME,
      }
    );

    expect(uploadStorageApi.addStorage).toHaveBeenCalledTimes(1);
    expect(uploadStorageApi.addStorage).toHaveBeenCalledWith(
      config.FILE_NAME,
      "mockReadStream"
    );

    expect(sourceFilesApi.listProjectBranches).toHaveBeenCalledTimes(1);
    expect(sourceFilesApi.listProjectBranches).toHaveBeenCalledWith(
      config.CROWDIN_PROJECT_ID,
      "mock-branch-name"
    );
  });

  it("downloads messages of the specified branch from Crowdin", async () => {
    expect.assertions(1);
    await collect(mockGlob);
    await download({
      translationsFile: config.TRANSLATIONS_FILE,
      translationsDir: config.TRANSLATIONS_DIR,
      branchName: config.BRANCH_NAME,
      languages: config.CROWDIN_LANGUAGES,
      typescript: false,
    });
    const file = fs.readFileSync(`${config.TRANSLATIONS_DIR}/nl.js`).toString();
    expect(file).toMatchSnapshot();
  });
});
