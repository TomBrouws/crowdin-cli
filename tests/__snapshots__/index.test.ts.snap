// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handlers downloads messages of the specified branch from Crowdin 1`] = `
"// Auto generated file. Do no change. Go to Crowdin to update the translations and run './node_modules/.bin/mollie-crowdin download' to update this file.
export default {
  \\"mollie-crowdin-content\\": \\"Dit is een test\\",
  \\"mollie-crowdin-title\\": \\"Titel\\",
};
"
`;
